Create the rankings page on rails!

AlbertoPdRF opened the pull request:
EDIT: it should be ready to go! :ship: 

I think I've come a long way already, but there are some things that are preventing me to finish with this and I figured out it would be better to discuss them while others are able to see the code.

Problems/questions I'm having:

1. I can't seem to get the JavaScript for the region selector to work :disappointed: I found the `request.params.merge(param: param)` to work quite well for the other selectors, but I don't know if that's what I should be using.
2. How can I test if what I wrote to support the old links work? I'm pretty sure it doesn't though :stuck_out_tongue: 
3. Are more tests necessary? If yes, of what kind?
4. This is quite nit-picky but... how should the SQL queries be indented?

Besides, I still have to find a clever way of crafting [the "By Region" table](https://www.worldcubeassociation.org/results/events.php?eventId=333&regionId=&years=&show=By%2BRegion&single=Single). I might come up with more questions regarding this later.

I'm not sure if I'm forgetting something, but with the problems/questions above and the code itself I think there's enough to start a conversation! :laughing:

Mollerz commented:
Something to add that I remember seeing an issue before but I can no longer find... It would be good to show WR, OcR, AfR, SAR, NAR, ER, AsR, or NR next to the time in the rankings list to show how the time stacks up against the rest of the world. 

Another thing that might be nice, I noticed on your first image you posted you had the flag of the person to show which country they are from. It might be a fun idea to also include the flag of the country next to the _competition_ to show where it was held as well! Especially if it shows WCA World Championship or a local city name, it is hardly ever obvious where the competition is held.

AlbertoPdRF commented:
> Something to add that I remember seeing an issue before but I can no longer find... It would be good to show WR, OcR, AfR, SAR, NAR, ER, AsR, or NR next to the time in the rankings list to show how the time stacks up against the rest of the world.

I would rather leave this for later, unless it's really trivial and people see it as necessary, and just focus on porting everything "as is." There are other suggested additions to the tables (some are on #301 itself,) and probably there's a better way to compute some of the tables (at least for some given combinations of parameters) than just porting and slightly modifying the SQL queries from the PHP code, but once we have everything in RoR we can better think how to improve it.

> Another thing that might be nice, I noticed on your first image you posted you had the flag of the person to show which country they are from. It might be a fun idea to also include the flag of the country next to the competition to show where it was held as well! Especially if it shows WCA World Championship or a local city name, it is hardly ever obvious where the competition is held.

See:

![image](https://user-images.githubusercontent.com/26309166/62469147-f59a8500-b797-11e9-95cb-e383bd0e3f86.png)

I'm fine keeping this, unless people feel like it looks a bit too overloaded of flags.

Mollerz commented:
Fair point on the record marking. I'll leave it up to others thoughts on the flags, personally I like it!

viroulep commented:
@tussosedan I know you expressed interest in porting stuff to rails, so maybe you'd like to check this out or help review this ;)

tussosedan commented:
@viroulep definitely interested! I'm not there yet though, need to find time to start with the basics :) It's a bit crazy now with the kids' summer vacation, but I will get to it for sure.

AlbertoPdRF commented on WcaOnRails/app/assets/stylesheets/results.scss:
```diff
@@ -0,0 +1,39 @@
+      padding: 5px 10px;
+      font-size: 13px;
+    }
+  }
```
Actually, I think only the `#results-query-form` is really necessary, so I think I'm going to get rid of this whole block.

AlbertoPdRF commented on WcaOnRails/app/assets/stylesheets/wca.scss:
```diff
@@ -350,6 +364,10 @@ textarea {
     }
   }
+
+  .active {
```
Agreed! I changed it to `a.active`.

AlbertoPdRF commented on WcaOnRails/app/helpers/results_helper.rb:
```diff
@@ -30,7 +30,7 @@ def historical_pb_markers(results)
 
   def link_to_ranking(event_id, type, &block)
-    url = "/results/events.php?eventId=#{event_id}&#{type}=true"
+    url = "/results/rankings/#{event_id}/#{type}"
```
`results_rankings_path(event_id, type)` doesn't seem to work, it's giving URLs like `results/rankings.event_id`. I've read that this is because you can't pass parameters like this to the `index` action. Maybe I could rename the action to `rankings` and try?

Another option is using `results_rankings_path(event_id: event_id, type: type)`, but it produces URLs like `/results/rankings?event_id=event_id&type=type`, which don't work either :slightly_frowning_face: 

The last option I can think of is leaving the helper (but as as an application helper) and use it for all the links to rankings.

AlbertoPdRF replied:
Funnily enough, `results_rankings_path(event_id, type)` works perfectly with the changes from 07e556770fe1a23609148917668b78f93a6ec149 to be able to test this out on staging :confused:

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+          value, countryId, year, month, day, personName
+      SQL
+    else
+      flash[:danger] = t(".unknown_show")
```
I'm happy to go with the majority here!

AlbertoPdRF commented on chef/site-cookbooks/wca/templates/worldcubeassociation.org.conf.erb:
```diff
@@ -147,6 +147,41 @@ server {
+    }
+
+    return 302 /competitions?$rails_args;
+  }
```
Sure thing, I'll try! (Together with the `location ~ ^\/results(?!\/rankings)` change below as discussed on Slack.)

AlbertoPdRF replied:
Argh, I still can't test this on staging :cry: 

![image](https://user-images.githubusercontent.com/26309166/62712502-40680700-b9fb-11e9-8352-b7a1ea0a721d.png)

AlbertoPdRF replied:
After chatting with @jfly, we will take care of this (and dropping the PHP code) on a separate PR, sticking to the `new-results` URL for now so we can test the code on the real production server.

jonatanklosko replied:
Ok, so I was playing around with that on the staging and finally got this to work. Ended up with the following configuration above `location /results`:
```nginx
# Redirect events.php to /results/rankings
location /results/events.php {
  set $rails_path "333";
  set $rails_args "";

  if ($arg_eventId != "") {
    set $rails_path $arg_eventId;
  }

  if ($arg_single != "") {
    set $rails_path "${rails_path}/single";
  }
  if ($arg_average != "") {
    set $rails_path "${rails_path}/average";
  }

  if ($arg_regionId != "") {
    set $rails_args "${rails_args}&region=${arg_regionId}";
  }

  if ($arg_years != "") {
    set $rails_args "${rails_args}&years=${arg_years}";
  }

  if ($arg_show != "") {
    set $rails_args "${rails_args}&show=${arg_show}";
  }

  return 302 /results/rankings/$rails_path?$rails_args;
}

location /results/rankings {
  try_files $uri @app;
}
```

AlbertoPdRF replied:
Nice! I will add this to the other PR when this one is merged 🙂

AlbertoPdRF commented on WcaOnRails/app/views/results/_index_form.html.erb:
```diff
@@ -0,0 +1,120 @@
+<div class="form-group">
+  <%= label_tag(t("results.index.form_elements.show_selector.show")) %>
+  <div id="show" class="btn-group">
+    <div class="btn-group">
```
I think yes, as I could only have one dropdown menu per button group. I'm leaving this conversation as unresolved for now just in case someone has a better solution!

AlbertoPdRF commented on WcaOnRails/app/views/results/index.html.erb:
```diff
@@ -0,0 +1,19 @@
+<div class="container">
+  <h1><%= yield(:title) %></h1>
+
+  <%= form_tag results_rankings_path, method: :get, remote: true, class: "results-select form-inline", id: "results-query-form" do %>
```
At the moment yes, I'm using plain links for everything, although I kept this because:

1. I didn't know if we would want to keep the direct links or investigate other possible solutions.
2. The regions selector is not working at the moment, and I still don't know how to make it work :stuck_out_tongue: I could turn it into a dropdown menu with `link_to`s as the others if you think that's a good option.

jonatanklosko replied:
I think links are just fine in this case. For the regions selector you can either try turning options into links, or add a piece of JS that simply changes `window.location` depending on the selected option value. Either way, we don't use a form for that.

AlbertoPdRF replied:
I added a piece of JS to take care of that and removed everything related to the form :slightly_smiling_face:

AlbertoPdRF commented on WcaOnRails/app/assets/javascripts/results.js:
```diff
@@ -0,0 +1,20 @@
+      });
+    }, 0);
+  });
+});
```
Nah, I left this just in case (same as above.)

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+    if params[:regionId].present?
+      params[:region_id] = params[:regionId]
+      params[:regionId] = nil
+    end
```
Makes sense, I'm removing this!

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+      if params[:years] == "all"
+        params[:years] = "all years"
+      elsif params[:years].include?("+")
+        params[:years].sub("+", " ")
```
That's nice to know, I've removed the `.sub("+", " ")`s :slightly_smiling_face:

jonatanklosko replied:
Ah sorry, apparently it is necessary in this case. Looking at the URLs I found out they include `%2B`, which is the escaped version of `+`, meaning we end up with an actual `+`! (same goes for `:show`)

AlbertoPdRF replied:
But actually, I've been trying on Chrome with both `+` and space on the URLs and they both worked just fine. Is it possible that it doesn't work on other browsers? Anyway, it's not a problem to add this back! 🙂

AlbertoPdRF replied:
After offline discussion, I've amended the PR so this is handled properly.

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+    if params[:show].present?
+      if params[:show].include?("All")
+        # We are not supporting the all option anymore!
+        params[:show] = "100 persons"
```
Makes sense, I'm changing this.

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+    @region_param = params[:region]
+    @type_param = params[:type]
+    years_param = params[:years]
+    show_param = params[:show]
```
I wasn't sure of this, so I'm happy to get rid of those and just use the `params[:param]`.

AlbertoPdRF commented on WcaOnRails/app/views/results/_index_table.html.erb:
```diff
@@ -0,0 +1,46 @@
+    <% previous_value = 0 %>
+    <% previous_rank = 0 %>
+    <% @rows.each_with_index do |row, i| %>
+      <% result = LightResult.new(row, Country.c_find(row["countryId"]), Format.c_find(row["formatId"]), RoundType.c_find(row["roundTypeId"]), Event.c_find(@event_param)) %>
```
I wasn't using `LightResult` until I discovered the `solve_tds_for_result` method on the results helper and that I couldn't use it without a result. That's pretty handy, so I would prefer to leave it that way :stuck_out_tongue: 

For the other idea, you mean changing that [here](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/models/light_result.rb#L42-L45), right? I'm fine with it if that's what you mean.

jonatanklosko replied:
Yeah, that's what I had in mind =)

AlbertoPdRF replied:
Ok, done!

AlbertoPdRF commented on WcaOnRails/app/views/results/_index_by_region_table.html.erb:
```diff
@@ -0,0 +1,62 @@
+        <tr>
+          <% (1..columns_number).each do %>
+            <td class="blank-row"><br></td>
+          <% end %>
```
Ups, that's right!

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+    end
+
+    @change_to_continents = world_rows.length
+    @change_to_countries = @change_to_continents + continents_rows.length
```
Sounds so much natural, I'm updating it.

AlbertoPdRF commented on WcaOnRails/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,279 @@
+    capitalized_type_param = @type_param.capitalize
+
+    event_condition = "AND eventId = '#{@event_param}'"
+    event_condition_without_and = "eventId = '#{@event_param}'"
```
I'll wait for further input on how to take care of this properly :slightly_smiling_face:

jfly replied:
I think there are two related, but different things to consider:

1. making sure that the user provided event_param is a valid event id.
2. making sure that the thing we put in the sql query cannot possibly do any harm.

Right now, doing 1) is "good enough", because our `eventId`s are "sql safe" (they don't contain any special characters). That said, I would rest easier if we *also* have the event param go through some sort of sanitization.

1) should be pretty straightforward to do

As for how to do 2)... It looks like [exec_query takes an optional `binds` parameter](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-exec_query), I think you can use that to replace `?`s in your SQL string with safely interpolated values. I'm having trouble finding decent documentation on this, unfortunately. It seems like the functionality in part depends on your database driver (in our case, that's the `mysql2` gem).

Urg, I just poked around for a while and couldn't figure out how to get the binds parameter to do anything useful. I eventually stumbled across this post: https://github.com/rails/rails/issues/20498, and it sounds like this just won't work with ActiveRecord + mysql2...

I'm not sure what the best approach is here. Some ideas for future investigation:

1. Find an escape method somewhere and try to remember to call it for every bit of string interpolation we do.
   - Pros: Straightforward
   - Cons: Need to find this method somewhere. Hopefully there's on buried in ActiveRecord somewhere. Also need to remember to use this everytime we do some string interpolation, which is just asking for trouble.
2. Use the `mysql2` gem directly, rather than going through ActiveRecord. `mysql2` looks like it supports [prepared statements](https://github.com/brianmario/mysql2#usage), which should let us do this in a safe way.
   - Cons: This ties us more closely to `mysql2`, which would make moving to a different database adapter (or even a different database!) more difficult in the future.
3. I don't know if this is possible, but: try to build an actual ActiveRecord model for this query. I get that we're not querying from a real table, but maybe AR supports querying from a subselect? If not, maybe we could trick AR by setting up a mysql view?

Of all these options, 3) feels like the most work, but also possibly the best approach. We've had some luck with using mysql views to trick AR before (see the `rails_persons` view for an example). Definitely curious what @jonatanklosko and @viroulep think, though.

viroulep replied:
Just like `competition_from_params` do a "find" on the competition model, I'd just do `Event.c_find!(...)` to make sure the page gives a 500 on an invalid event id.

AlbertoPdRF replied:
I've added the `Event.c_find!(params[:event_id])` to make sure the page crashes on an invalid `:event_id`. I'm not sure if something like this is necessary for other params too?

- `:type`:
  - it's used to get `value`, which will be either `best` or `average` (there's no other option,) which then is interpolated on some of the queries
  - it's capitalized and interpolated on some of the queries between `Concise` and `Results`, without spaces
- `:region`:
  - it's used to either find a continent or a country. If none is found, no interpolation is made.
- `:show`:
  - it's used to decide which query to run
  - it's splitted and the first part is directly interpolated on all but one query after a `LIMIT `
- `:years`
  - it's splitted; its first part is used to decide if an `=` or `<=` sign will be interpolated on the queries together with the splitted second part, after a `AND year `. If it doesn't match exactly, no interpolation happens.

Do you think some of them need to be (further) sanitized?

jonatanklosko replied:
So looks like `year` and `show` are not really sanitized, would make sense to just parse them to ints.

AlbertoPdRF replied:
I've added a `.to_i` after splitting them.

jfly replied:
> `:type` ... will be either best or average (there's no other option,)

As I'm reading the code, it looks like `type` comes directly from the user, gets capitalized, and then injected directly into a SQL string. Why do you say it will be either "best" or "average"?

I agree that the use of `:region` and show `:show` look ok as they are now. Adding a `.to_i` for `show` seems good.

The current use of `:years` does *not* look safe to me. Adding a `.to_i` does feel like it would make things safe.

AlbertoPdRF replied:
> As I'm reading the code, it looks like type comes directly from the user, gets capitalized, and then injected directly into a SQL string. Why do you say it will be either "best" or "average"?

`type` is used for two things (note the two sub-points on my previous comment.) The first time it's used is [here](https://github.com/thewca/worldcubeassociation.org/pull/4394/files#diff-09cacc147fbd254009efaa431d0cc8b2R17-R18), hence that comment that you quoted. The second one is when it's directly interpolated. Any suggestion on how to improve that?

> The current use of :years does not look safe to me. Adding a .to_i does feel like it would make things safe.

I had already added it, see [here](https://github.com/thewca/worldcubeassociation.org/pull/4394/files#diff-09cacc147fbd254009efaa431d0cc8b2R47)

AlbertoPdRF replied:
Maybe something like:

```
if !@types.include?(params[:type])
  flash[:danger] = t(".unknown_type")
  return redirect_to results_rankings_path
end
```

would work for the `type` param sanitization?

AlbertoPdRF commented:
After quickly chatting with @jfly, I pushed 07e556770fe1a23609148917668b78f93a6ec149 to test the code on staging. With that change, I could try and see that everything works as expected. I didn't leave this deployed to staging though, given that https://github.com/thewca/worldcubeassociation.org/pull/4394#discussion_r311974383 still needs to be taken care of.

viroulep commented:
There are a lot of plain SQL in the PR currently (ported from the php obviously :)), and just like Lars and Tim mentioned in previous "results work" [here](https://github.com/thewca/worldcubeassociation.org/pull/1075#issuecomment-271100416), I think we should make some effort to get what we want through ActiveRecord.

Note: I did not actually look into what the queries were doing, so maybe we *have* to write them manually.

Good luck getting everything to work @AlbertoPdRF, that's definitely a big task you're taking on!

AlbertoPdRF commented:
I actually started working on this by trying to get things through ActiveRecord, but after finding some difficulties and chatting with @jfly, he recommended me to use the queries directly and said that we could try to improve things when the page is ported.

> Good luck getting everything to work @AlbertoPdRF, that's definitely a big task you're taking on!

Thanks :smile_cat:

jonatanklosko commented on WcaOnRails/app/views/results/_index_by_region_table.html.erb:
```diff
@@ -0,0 +1,60 @@
+  <tbody>
+    <% shown = {} %>
+    <% @rows_to_display.each_with_index do |row, i| %>
+      <% result = LightResult.new(row, row["countryId"], row["formatId"], row["roundTypeId"], params[:event_id]) %>
```
All the other arguments are included in `row`, so my idea was to just pass the row ^^

AlbertoPdRF replied:
Right, that makes sense :slightly_smiling_face: I've changed it already, adding `result.eventId       eventId,` to one of the queries that was missing it!

AlbertoPdRF commented:
I think I've addressed all the concerns. After chatting with @jfly, we have decided that this PR should be merged using the `new-results` URL and without removing the old PHP code so everything can be tested on the real production server. Another PR will follow with the changes that I dropped from this one to make the switch.

jonatanklosko commented on WcaOnRails/app/views/results/_rankings_selector.html.erb:
```diff
@@ -0,0 +1,144 @@
+        }
+      }
+      window.location.href = url + hash;
+    });
```
I believe we have some jQuery plugins for managing params, this could be something like:
```js
var params = $.getUrlParams();
params.region = $(this).val();
window.location.search = $.param(params);
```

AlbertoPdRF replied:
That's so much better, I've updated it already! :slightly_smiling_face:

AlbertoPdRF commented:
Thanks guys for all the help! I'm really excited for this to go live :smile:
